pycheribuild/projects/go.py

```diff
@@ -0,0 +1,90 @@
+# ("CTSRD"), as part of the DARPA CRASH research programme.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+from pathlib import Path
+import shutil
+
+import os
+import sys
+
+from .project import *
+from ..utils import *
+from ..config.loader import ComputedDefaultValue
+
+
+class BuildGo(Project):
+    githubBaseUrl = "https://github.com/CTSRD-CHERI/"
+    repository = GitRepository(githubBaseUrl + "freebsd-mips-go.git")
+    no_default_sysroot = None
+    appendCheriBitsToBuildDir = True
+    skip_cheri_symlinks = True
+    defaultInstallDir = Project._installToSDK
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+        # It does not seem possible to change this in the go build scripts (easily).
+        self.makeDir = self.sourceDir / "src"
+        self.binDir = self.sourceDir / "bin"
+        self.pkgDir = self.sourceDir / "pkg"
+        self.gorootDir = self.installDir / "go"
+
+    def buildDirForTarget(self, target: CrossCompileTarget):
+        return self.sourceDir / "pkg"
+    
+    def process(self):
+        if not self.config.skipUpdate:
+            self.update()
+        env = {
+            "GOROOT_FINAL": self.gorootDir,
+        }
+        cmd = "bash make.bash".split()
+        # cmd = "bash all.bash".split()
+        runCmd(cmd, cwd=self.makeDir, env=env)
+        self.install()
+
+    def clean(self) -> ThreadJoiner:
+        runCmd("bash clean.bash".split(), cwd=self.makeDir)
+        return ThreadJoiner(None)
+
+    def install(self):
+        # Move bin and pkg to goroot and link src dir
+        shutil.rmtree(self.gorootDir, ignore_errors=True)
+        os.makedirs(self.gorootDir)
+        shutil.copytree(self.binDir, self.gorootDir / "bin")
```

Comment:
This breaks the --pretend mode. Probably best to add a `if self.config.pretend: return` before the copytree.

Alternatively I added wrappers to project for unlink (`self.deleteFile()`) and `self.createSymlink()`.

I haven't got a wrapper for copytree yet (only for single files, `self.installFile()`) but I could always add that.